presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestMaterializedViewQueryOptimizer.java

```diff
@@ -464,436 +478,724 @@ public void testWithJoinTables()
+        String viewSql2 = subquery2;
+
+        String baseQuerySql = format("SELECT GREATEST(count_a1, count_a2) AS bigcount FROM (%s) s1 INNER JOIN (%s) s2 ON s1.b = s2.c", subquery1, subquery2);
+
+        String expectedRewrittenSql = format("SELECT GREATEST(count_a1, count_a2) AS bigcount FROM " +
+                "(SELECT SUM(count_a1) AS count_a1, b FROM %s GROUP BY b) s1 " +
+                "INNER JOIN " +
+                "(SELECT SUM(count_a2) AS count_a2, c FROM %s GROUP BY c) s2 " +
+                "ON s1.b = s2.c",
+                VIEW_1, VIEW_2);
+
+        assertOptimizedQuery(baseQuerySql, expectedRewrittenSql, viewSql1, viewSql2);
+
+        subquery1 = format("SELECT MIN(b) AS min_b FROM %s", BASE_TABLE_1);
+        subquery2 = format("SELECT MIN(c) AS min_c FROM %s", BASE_TABLE_2);
+
+        viewSql1 = subquery1;
+        viewSql2 = subquery2;
+
+        baseQuerySql = format("SELECT min_b AS miny_b FROM (%s) s1 INNER JOIN (%s) s2 ON s1.min_b = s2.min_c", subquery1, subquery2);
+
+        expectedRewrittenSql = format("SELECT min_b AS miny_b FROM " +
+                        "(SELECT MIN(min_b) AS min_b FROM %s) s1 \n" +
+                        "INNER JOIN " +
+                        "(SELECT MIN(min_c) AS min_c FROM %s) s2 \n" +
+                        "ON s1.min_b = s2.min_c",
+                VIEW_1, VIEW_2);
+
+        assertOptimizedQuery(baseQuerySql, expectedRewrittenSql, viewSql1, viewSql2);
+    }
+
+    @Test
+    public void subqueryAggregationUnsupportedFunction()
+    {
+        String subquery1 = format("SELECT GEOMETRIC_MEAN(b) AS mean_b FROM %s", BASE_TABLE_1);
+        String subquery2 = format("SELECT GEOMETRIC_MEAN(c) AS mean_c FROM %s", BASE_TABLE_2);
+
+        String viewSql1 = subquery1;
+        String viewSql2 = subquery2;
+
+        String baseQuerySql = format("SELECT mean_b AS meany_b FROM (%s) s1 INNER JOIN (%s) s2 ON s1.mean_b = s2.mean_c", subquery1, subquery2);
+
+        assertOptimizedQuery(baseQuerySql, baseQuerySql, viewSql1, viewSql2);
+    }
+
+    @Test
+    public void subqueryAggregationSomeSupportedFunctions()
+    {
+        String subquery1 = format("SELECT min(b) AS min_b FROM %s", BASE_TABLE_1);
+        String subquery2 = format("SELECT GEOMETRIC_MEAN(c) AS mean_c FROM %s", BASE_TABLE_2);
+
+        String viewSql1 = subquery1;
+        String viewSql2 = subquery2;
+
+        String baseQuerySql = format("SELECT min_b AS miny_b FROM (%s) s1 INNER JOIN (%s) s2 ON s1.min_b = s2.mean_c", subquery1, subquery2);
+
+        String expectedRewrittenSql = format("SELECT min_b AS miny_b FROM " +
+                        "(SELECT MIN(min_b) AS min_b FROM %s) s1 " +
+                        "INNER JOIN " +
+                        "(%s) s2 " +
+                        "ON s1.min_b = s2.mean_c",
+                VIEW_1, subquery2);
+
+        assertOptimizedQuery(baseQuerySql, expectedRewrittenSql, viewSql1, viewSql2);
+    }
+
+    @Test
+    public void subqueryJoinAggregates()
```

Comment:
Can you make some of subqueries slightly more complex? Like some more group bys and wheres than just what materialized view has. You can just copy those rewrites from above tests